### PR TITLE
Add counter filter

### PR DESCRIPTION
**Pull request by keilr**
##### SUMMARY

A filter to count (hashable) elements in a sequence:

  - count the number of occurrences of each character in a string
  - count occurrences of each element in a list

##### ISSUE TYPE
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
filter

### CODE REVIEW

**Comment by keilr** [comment]
the failed CI jobs don't seem to be related to the new filter plugin

**Comment by felixfontein** [comment]
Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to) (please note the special syntax for new filter plugins) and documentation in https://github.com/ansible-collections/community.general/blob/main/docs/docsite/rst/filter_guide.rst ? Thanks!

**Comment by keilr** [comment]
@felixfontein thanks a lot for the review! I'll add the missing changelog and documentation shortly.

What do you think of renaming the plugin to `counter` (name of the [python class](https://docs.python.org/3/library/collections.html#collections.Counter) used within the plugin)? There's a Jinja2 default filter `count` (alias for `length`), which could be misleading, but as most users probably use fully qualified names (`community.general.counter`), this shouldn't confuse.

**Comment by felixfontein** [comment]
> @felixfontein thanks a lot for the review! I'll add the missing changelog and documentation shortly.
> 
> What do you think of renaming the plugin to `counter` (name of the [python class](https://docs.python.org/3/library/collections.html#collections.Counter) used within the plugin)? There's a Jinja2 default filter `count` (alias for `length`), which could be misleading, but as most users probably use fully qualified names (`community.general.counter`), this shouldn't confuse.

There is no way to use this plugin without FQCN, so there's even less reason for confusion. Your name suggestion sounds good to me.

File: plugins/filter/counter.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+def counter(sequence):
+    ''' Count elements in a sequence. Returns dict with count result. '''
+    if not isinstance(sequence, Sequence):
+        raise AnsibleFilterError('First argument for counter must be a sequence (str or list). %s is %s' %
+                                 (sequence, type(sequence)))
+
+    for element in sequence:
+        try:
+            hash(element)
```
Comment (felixfontein) [inline plugins/filter/counter.py:21]: Since `Counter()` already does this: how about adding a `try`/`except TypeError` around line 26 instead?

Comment (keilr) [reply plugins/filter/counter.py]: Indeed, that's much nicer. Fixed with https://github.com/ansible-collections/community.general/pull/3921/commits/cc77c72b4398245baac8e8ded819101a8dbb886d

**Comment by keilr** [comment]
@felixfontein doc moved to 'Abstract transformations' chapter, see https://github.com/ansible-collections/community.general/pull/3921/commits/4be5b47c846c4335d33b959777f931e44d796ce4

I'll squash the commits after getting the "OK" from you.

**Comment by felixfontein** [comment]
> I'll squash the commits after getting the "OK" from you.

Please don't squash the commits; we always squash on merge (see also https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#open-pull-requests).

File: docs/docsite/rst/filter_guide.rst [hunk 2]
```diff
@@ -297,6 +297,84 @@ This produces:
...
+          sdf:
+            adapter: scsi_3
+          sdg:
+            adapter: scsi_3
+
+This produces:
+
+.. code-block:: ansible-output
+
+    TASK [Get a SCSI controller id with less than 4 disks and choose the one with least disks]
```
Comment (felixfontein) [inline docs/docsite/rst/filter_guide.rst:371]: This message is different from the `name` of the task above :)

**Review by felixfontein — COMMENTED** [review]
Besides this, looks good!

File: docs/docsite/rst/filter_guide.rst (hunk 2 above)
Comment (keilr) [reply docs/docsite/rst/filter_guide.rst]: thanks :) fixed with 04df7c1

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@keilr thanks for your contribution!

**Pull request merged**